### ISSUE TITLE
About MultiHeadAttention's split method

### ISSUE THREAD

**Issue by 0x6f6f** [hyunwoongko/transformer#21]
### Hey bro I find it confusing about ".contiguous()" usage in this class.

I'm a beginner about this field, and I just learned the purpose of using the method is make tensors save in contiguous space and increase effeciency. 

But in this code:
`class MultiHeadAttention(nn.Module):

    def __init__(self, d_model, n_head):
        super(MultiHeadAttention, self).__init__()
        self.n_head = n_head
        self.attention = ScaleDotProductAttention()
        self.w_q = nn.Linear(d_model, d_model)
        self.w_k = nn.Linear(d_model, d_model)
        self.w_v = nn.Linear(d_model, d_model)
        self.w_concat = nn.Linear(d_model, d_model)

    def forward(self, q, k, v, mask=None):
        # 1. dot product with weight matrices
        q, k, v = self.w_q(q), self.w_k(k), self.w_v(v)

        # 2. split tensor by number of heads
        q, k, v = self.split(q), self.split(k), self.split(v)

        # 3. do scale dot product to compute similarity
        out, attention = self.attention(q, k, v, mask=mask)
        
        # 4. concat and pass to linear layer
        out = self.concat(out)
        out = self.w_concat(out)

        # 5. visualize attention map
        # TODO : we should implement visualization

        return out

    def split(self, tensor):
        """
        split tensor by number of head

        :param tensor: [batch_size, length, d_model]
        :return: [batch_size, head, length, d_tensor]
        """
        batch_size, length, d_model = tensor.size()

        d_tensor = d_model // self.n_head
        tensor = tensor.view(batch_size, length, self.n_head, d_tensor).transpose(1, 2)
				
        # it is similar with group convolution (split by number of heads)

        return tensor

    def concat(self, tensor):
        """
        inverse function of self.split(tensor : torch.Tensor)

        :param tensor: [batch_size, head, length, d_tensor]
        :return: [batch_size, length, d_model]
        """
        batch_size, head, length, d_tensor = tensor.size()
        d_model = head * d_tensor

        tensor = tensor.transpose(1, 2).contiguous().view(batch_size, length, d_model)
        return tensor`

I find that we din't use .contiguous() in split method while using that method in concat method.

Is there any special reason? I'd appreciate it a lot if anyone could answer me.

**Comment by zhangjingxian1998** [comment]
It's all depend on wether you want to use ```.view()``` to reshape the shape of the tensor.
If you want, you must to use ```.contiguous()```
By the way, there is another method ```.reshape()``` to achive the same result of ```.view()```, but the ```.reshape()``` don't need to use  ```.contiguous()``` before it.

**Comment by zhangjingxian1998** [comment]
for example
a input tensor ```A``` have shape ```[8,50,768]```, after split, the shape is ```[8,12,50,64]``` if there is 12 heads.
```
A                        # [8,50,768]
A = split(A)             # [8,12,50,64]
A.view(8, 12, 50*64)     # wrong, because uncontiguous
A.view(8, 12, 50, 32, 2) # right, because it is contiguous in 64
```

**Comment by 0x6f6f** [comment]
> for example a input tensor `A` have shape `[8,50,768]`, after split, the shape is `[8,12,50,64]` if there is 12 heads.
> 
> ```
> A                        # [8,50,768]
> A = split(A)             # [8,12,50,64]
> A.view(8, 12, 50*64)     # wrong, because uncontiguous
> A.view(8, 12, 50, 32, 2) # right, because it is contiguous in 64
> ```

thanks a lot, that's helpful.
